### ISSUE TITLE
Difference in evaluator

### ISSUE THREAD

**Issue by yolor2** [ultralytics/yolov5#8123]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello @glenn-jocher , when i am testing my test data using test.py  using confidence threshold of 0.01 then the mAP of both from yolov5 evaluator i.e mAP50_95 is around 59.6% while using COCO evaluator it is mAP50_95 is 59.5%. which is nearly the same. But when i am increasing my confidence threshold to 0.4 the mAP using yolov5 evaluator becomes 65% and from coco evaluator mAP50_95 is 56.4%.
Could you please help me differentiate between this two evaluator. Or at least how the YOLOv5 evaluator is different from COCO. Could you please explain why increasing confidence threshold increases mAP using YOLOv5 evaluator and mAP decreases for COCO evaluator. Please try to explain and i am really confused.

**Comment by glenn-jocher** [comment]
@yolor2 higher confidence are invalid for metrics computation as the warning clearly states
https://github.com/ultralytics/yolov5/blob/2dd3db0050cd228e7a7ca3ff72ab7e3f34ea64d7/val.py#L365-L366

**Comment by yolor2** [comment]
@glenn-jocher , why is this case. As i am validating the trained weights on my unseen test dataset and check mAP of the model on test set. Now if i use less confidence threshold then there would be more false positives.
But why using detect.py i am using higher confidence threshold, now over there if i use low confidence threshold then multiple bounding boxes are drawn and sometimes  for one object both the classses gets allocated( as i am training for 2 classes).
Moreover, why does there is a difference in COCO mAP and YOLOv5 evaluator as we increase confidence threshold. YOLOv5 evalutor shows increase in mAP while coco evaluator shows decreasein mAP with increase in confidence threshold upto 0.4
Could you please enlighten me more.

**Comment by glenn-jocher** [comment]
@yolor2 👋 Hello, thanks for asking about the differences between [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py), [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) and [val.py](https://github.com/ultralytics/yolov5/blob/master/val.py) in YOLOv5 🚀.

These 3 files are designed for different purposes and utilize different dataloaders with different settings. train.py dataloaders are designed for a speed-accuracy compromise, val.py is designed to obtain the best mAP on a validation dataset, and detect.py is designed for best real-world inference results. A few important aspects of each:


## train.py

- **trainloader** LoadImagesAndLabels(): designed to load train dataset images and labels. Augmentation capable and enabled. https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/train.py#L210-L213 **val_loader** LoadImagesAndLabels(): designed to load val dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/train.py#L220-L223
- image size: `640`
- rectangular inference: `False`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `None`

## val.py

- **dataloader** LoadImagesAndLabels(): designed to load train, val, test dataset images and labels. Augmentation capable but disabled. 
- https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/val.py#L152-L153
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `0.5 * maximum stride`


## detect.py

- **dataloaders** (multiple): designed for loading multiple types of media (images, videos, globs, directories, streams). https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/detect.py#L120-L128
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L152): `0.25` 
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L153): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/utils/general.py#L415): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L45-L54): `None`


## YOLOv5 PyTorch Hub Inference

YOLOv5 PyTorch Hub models are`AutoShape()` instances used for image loading, preprocessing, inference and NMS. For more info see [YOLOv5 PyTorch Hub Tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)
https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/models/common.py#L276-L282

- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.25 `
- [iou threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L288): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `None`

Good luck 🍀 and let us know if you have any other questions!

**Comment by yolor2** [comment]
@glenn-jocher Thank you for your reply. I understood this files, but suppose if i want to obtain the mAP of test set, to check how the trained weights perform on test set.At that point how to obtain the mAP of the already labelled test set.
In short what is the difference in confidence threshold of val.py and detect.py. I mean not numerically but conceptually. Because i want to showcase the mAP of test set 
Thank you and really need your help to understand the concept.

**Comment by glenn-jocher** [comment]
@yolor2 for mAP on a labelled test set:
```
python val.py --task test
```

**Comment by yolor2** [comment]
@glenn-jocher then what would be the confidence threshold used for this purpose. If we you could explain me the difference in confidence threshold used for val.py and detect.py
and why can not be the confidence threshold greater than 0.001 for evaluation of metrics.

**Comment by glenn-jocher** [comment]
@yolor2 val.py should always use default confidence threshold of 0.001
